### PR TITLE
Add option to add explorer context menu in Windows installer

### PR DESCRIPTION
**Pull request by bergmeister**
## PR Summary
A new dialogue is added in the Windows installer to offer the option of adding explorer.exe context menus to open PowerShell in the current location as a normal shell or as administrator. The context menu entries are achiveven via registry entries and is available when right-clicking on:

- the background of explorer
- a folder in explorer
- a drive in explorer
- on the Desktop

Adding the new custom dialog required copy-pasting the WiX source code of [WixUI_InstallDir.wxs](https://github.com/wixtoolset/wix3/blob/7c8457fa3802858add8981b92996e19574e762da/src/ext/UIExtension/wixlib/WixUI_InstallDir.wxs) as a base as described in the official WiX documentation [here](http://wixtoolset.org/documentation/manual/v3/wixui/wixui_customizations.html).
The registry entries are quite standard but the approach for the context menu of opening PowerShell as an Administrator must be discussed:
The Windows registry has the limitation that it only offers a `runas/command` structure but not a `runas/proogram/command` structure, which means there can only be one main context entry to open PowerShell as an Admin. Since a user might have alerady configured this context menu, I decided that it is best to not stomp over it but rather open PowerShell normally but then elevate it. This of course has the disadvantage that it takes a bit longer and a window is visible for a brief moment. The only alternative that would allow to launch PowerShell elevated the first time is to have the 2 context menu in a submenu, see [here](https://blogs.msdn.microsoft.com/andrew_richards/2017/03/01/enhancing-the-open-command-prompt-here-shift-right-click-context-menu-experience/)
The registry keys contain the PowerShell version to cater for side by side installations of future versions.
Some stuff had to be created in a WiX fragment, which means that this XML section could be factored out into its own file (with the disadvantage of having to linkt this file then into the compilation process of candle,, light, etc.)
The best person to review is probably @SteveL-MSFT 

This is how it currently looks like: 
Installer: 
image
Explorer context menu:
image

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [?] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission) Maybe document the WiX parameter to enable users to use this feature as well when installing quietly?
    - [x] Issue filed - Issue link: https://github.com/PowerShell/PowerShell/issues/5012
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting) A test for this would require a UI automation framework like Ranorex or Selenium.
    - [NA] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

**Comment by SteveL-MSFT** [comment]
@bergmeister Thanks for doing this.  I think I prefer the submenu to starting a process to start elevated PowerShell.  cc @joeyaiello

**Comment by bergmeister** [comment]
Ok. Any preference on the naming of the menu? I am thinking of `PowerShell X.Y.Z` for the menu and then `Open here`\`Open as Administrator here for the 2 submenus.

**Comment by bergmeister** [comment]
@SteveL-MSFT I changed it to use submenus to properly elevate it. This is how it looks now (the logic whether the submenu pops out on the left or right hand side is handled by Windows depending on where one right-clicks):
image
Do we need to repeat the product name and version in the sub menu like the explorer shortcut for Windows PowerShell in Win10 does it?
image

**Comment by bergmeister** [comment]
NB: I also tweaked the tabbing index on the new installer page and to be able to use the spacebar to toggle the checkbox. Additionally, the commands also have accessibility shortcuts now (those are the ampersands in the commands), which means that the top context menu can be selected using the `P` key (or pressing it multiple times if other menus also have the same key configured) and `H` or `A` for the submenu (the latter is the Admin shell). I am done now and the PR can be reviewed again.

**Review by SteveL-MSFT — APPROVED** [review]
LGTM, built off your branch and works for me.  Great work!

**Comment by bergmeister** [comment]
Thanks. Do you think we should document the new property `EXPLORER_CONTEXT_MENU_OPENPOWERSHELL` for people who want to install the MSI quietly (`/q` option ot `msiexec.exe`)?

**Comment by SteveL-MSFT** [comment]
@bergmeister yes, we should document it.  However, since this msi is new, we don't have an existing doc to put it.  @sdwheeler can you suggest how we should document the property that can be passed to the msi installer for non-interactive (quiet) installs?

**Comment by sdwheeler** [comment]
@bergmeister @SteveL-MSFT Please create an issue for this in the PowerShell-Docs repo. It should probably be assigned to Joey. He is working on the V6 setup docs. Joey and I can work out how and where this should be documented. Also include all possible MSI properties that could be used so we can be complete.

**Comment by SteveL-MSFT** [comment]
Created https://github.com/PowerShell/PowerShell-Docs/issues/2007

**Comment by bergmeister** [comment]
@sdwheeler The new property is the only one as far as I am aware. But I thought that this would go only into 6.1 or this a candidate for merge to the release branch @SteveL-MSFT  ?

**Comment by SteveL-MSFT** [comment]
@bergmeister this is planned for 6.1

**Comment by iSazonov** [comment]
Currently we have a nuget package to set `PSModulePath` with Windows PowerShell module paths. It is not obvious. My request (for follow PR) is to add the option to the installer too.

**Comment by iSazonov** [comment]
>Do you think we should document the new property EXPLORER_CONTEXT_MENU_OPENPOWERSHELL 

Maybe `ADD_EXPLORER_CONTEXT_MENU_OPENPOWERSHELL`.

**Review by iSazonov — APPROVED** [review]

**Comment by bergmeister** [comment]
@iSazonov I applied your renaming suggestion. Please open an issue about your `PSModulePath` request with more details in it, I would be happy to help you by doing the UI/WiX bit of it in case you haven't worked much with WiX much yet.

**Comment by iSazonov** [comment]
@bergmeister Thanks!
I'll open a Issue about `PSModulePath` if @SteveL-MSFT approve.

File: assets/Product.wxs [hunk 1]
```diff
@@ -130,4 +179,105 @@
...
+
+      <Publish Dialog="ExitDialog" Control="Finish" Event="EndDialog" Value="Return" Order="999">1</Publish>
+
+      <Publish Dialog="WelcomeDlg" Control="Next" Event="NewDialog" Value="LicenseAgreementDlg">NOT Installed</Publish>
+      <Publish Dialog="WelcomeDlg" Control="Next" Event="NewDialog" Value="VerifyReadyDlg">Installed AND PATCH</Publish>
+
+      <Publish Dialog="LicenseAgreementDlg" Control="Back" Event="NewDialog" Value="WelcomeDlg">1</Publish>
+      <Publish Dialog="LicenseAgreementDlg" Control="Next" Event="NewDialog" Value="InstallDirDlg">LicenseAccepted = "1"</Publish>
+
+      <Publish Dialog="ExplorerContextMenuDialog" Control="Back" Event="NewDialog" Value="InstallDirDlg">1</Publish>
```
Comment (KalleOlaviNiemitalo) [inline assets/Product.wxs]: This implementation of the Back button looks OK, does not have bug <https://github.com/PowerShell/PowerShell/issues/12055> yet.

**Pull request merged**